crates/ruff_linter/src/rules/ruff/rules/sort_dunder_slots.rs

```diff
@@ -0,0 +1,272 @@
+        return;
+    };
+
+    let sort_classification = SortClassification::of_elements(&elts_analysis.elts, natord::compare);
+    if sort_classification.is_not_a_list_of_string_literals() || sort_classification.is_sorted() {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(
+        UnsortedDunderSlots {
+            class_name: class_name.to_string(),
+            class_variable: dunder_kind,
+        },
+        elts_analysis.range,
+    );
+
+    if let SortClassification::UnsortedAndMaybeFixable { items } = sort_classification {
+        if let Some(fix) = create_fix(&elts_analysis, &items, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+#[derive(Debug)]
+struct NodeAnalysis<'a> {
+    elts: Cow<'a, Vec<ast::Expr>>,
+    range: TextRange,
+    display_kind: DisplayKind<'a>,
+}
+
+impl<'a> NodeAnalysis<'a> {
+    fn new(node: &'a ast::Expr, dunder_kind: SpecialClassDunder) -> Option<Self> {
+        let result = match (dunder_kind, node) {
+            (_, ast::Expr::List(ast::ExprList { elts, range, .. })) => {
+                let display_kind = DisplayKind::Sequence(SequenceKind::List);
+                Self {
+                    elts: Cow::Borrowed(elts),
+                    range: *range,
+                    display_kind,
+                }
+            }
+            (_, ast::Expr::Tuple(tuple_node @ ast::ExprTuple { elts, range, .. })) => {
+                let display_kind = DisplayKind::Sequence(SequenceKind::Tuple(tuple_node));
+                Self {
+                    elts: Cow::Borrowed(elts),
+                    range: *range,
+                    display_kind,
+                }
+            }
+            (SpecialClassDunder::Slots, ast::Expr::Set(ast::ExprSet { elts, range })) => {
+                let display_kind = DisplayKind::Sequence(SequenceKind::Set);
+                Self {
+                    elts: Cow::Borrowed(elts),
+                    range: *range,
+                    display_kind,
+                }
+            }
+            (
+                SpecialClassDunder::Slots,
+                ast::Expr::Dict(ast::ExprDict {
+                    keys,
+                    values,
+                    range,
+                }),
+            ) => {
+                let mut narrowed_keys = Vec::with_capacity(values.len());
+                for key in keys {
+                    if let Some(key) = key {
+                        // This is somewhat unfortunate,
+                        // *but* only `__slots__` can be a dict out of
+                        // `__all__`, `__slots__` and `__match_args__`,
+                        // and even for `__slots__`, using a dict is very rare
+                        narrowed_keys.push(key.to_owned());
+                    } else {
+                        return None;
+                    }
+                }
```

Comment:
You could create a `Keys(&[Option<Expr>])` newtype where the `new` function returns `None` when any value is `None` and `Some(Self)` otherwise. 

The `Keys` implementation can than expose functions to access the elements: 

```rust
impl<'a> Keys<'a> {
    fn iter() -> impl Iterator<Item=&'a Expr> {
        // SAFETY: Safe because the constructor guarantees that all values are non-None. 
        self.0.iter().map(Option::unwrap)
    }

    ... 
}
```

But yeah, don't know if it's worth it. It probably depends on the boilerplate necessary.

Reply:
Oh, that's clever! Hey, why not. Doesn't seem _too_ hard to do.

Reply:
Hmm, okay. Harder than I thought. Tempted to leave this for now, since, as I say, this is an edge case that it's really unlikely we'd hit very often